OSDOCS-20033: Kueue 1.4 and DRA (#113996)
openshift/openshift-docs: ai_workloads/kueue/kueue-dra-integrating-dynamic-resource-allocation.adoc

:_mod-docs-content-type: ASSEMBLY
include::_attributes/common-attributes.adoc[]
[id="kueue-dra-integrating-dynamic-resource-allocation"]
= Integrating Dynamic Resource Allocation
:context: kueue-dra-integrating-dynamic-resource-allocation

toc::[]

[role="_abstract"]

> {kueue-name} Dynamic Resource Allocation (DRA) integration enables advanced management of specialized hardware resources like GPUs, FPGAs, and other accelerators within Kubernetes workload queuing. This integration allows for the reading and publishing of ResourceSlices, counter-based quota computation, and specific admission behaviors. 

Reviewer: ```suggestion
You can configure {kueue-name} to manage quota for workloads that use Dynamic Resource Allocation (DRA) to request GPUs. When DRA quota management is configured, {kueue-name} counts DRA device requests toward quota in the same way that it counts traditional resources such as CPU and memory.
```

Author: Replaced.

> DRA is a Kubernetes framework that manages specialized hardware resources such as GPUs with fine-grained control. Unlike traditional resource requests, DRA allows dynamic prioritization—allocating GPUs to high-priority AI training workloads during business hours, then reallocating them to cost-optimized batch jobs overnight.

Reviewer: ```suggestion
If DRA device quota is not configured, {kueue-name} does not account for GPU requests when admitting workloads, which can result in teams exceeding their GPU allocation.
```

Author: Replaced.

> You can validate partitionable devices support in {kueue-name} Dynamic Resource Allocation (DRA) integration, covering partition-aware quota, admission, and scheduling. Partitionable devices, such as NVIDIA MIG, allow graphics processing units (GPUs) to be dynamically subdivided into smaller allocations. {kueue-name} must correctly handle quota accounting for these mutually exclusive partition configurations.

Reviewer 2: It's good that you've mentioned Partitionable Devices here. Can you also mention something about Structured Parameters and Extended Resources? They are also "features" or "resources" provided by DRA.

Author: I added some text for these.

> 

Reviewer: L15-L25, not required.

Author: @sohankunkerkar Deleting lines 15-25 would leave the 2 bulleted items on lines 26-28 hanging. Should those be deleted too?

Reviewer: Are you talking about the first two points from the `Prerequisites`?

Author: No, the last two bullets, shown here: 
<img width="1476" height="532" alt="image" src="https://github.com/user-attachments/assets/b66eefb8-0675-46f1-9099-c2ab826faebe" />

Author: Rewrote this topic as per Sohan's input.

{kueue-name} DRA integration allows for:

* Admission validation with specific partition sizes and quota enforcement.

* Testing of mixed partition workloads on the same device and preemption across partition sizes.

* Verification of partition capacity reclaim after workload completion.

* Validation using `nvidia-dra-driver`.

.Prerequisites
* {kueue-name} is installed.

> * {product-title} running version 4.21 or later.

Reviewer 2: Partitionable Devices requires OCP 4.22+ (K8s 1.35 with DRAPartitionableDevices gate). The PD module itself says "4.22 or later" — contradicting this prerequisite. The prerequisite should either list both versions or note the PD exception.

Author: Added a Note: "To use partitionable devices, you need {product-title} 4.22 or later. "

> * A DRA driver installed in the cluster, for example, `nvidia-dra-driver` or `k8s-dra-driver-gpu`.

Reviewer 2: k8s-dra-driver-gpu is not a standard driver name as far as I know. The upstream example driver is dra-example-driver (which publishes as gpu.example.com) but, I think Sohan asked you to remove that.

Author: Removed `k8s-dra-driver-gpu`

> * A DRA driver installed in the cluster, for example, `nvidia-dra-driver` or `k8s-dra-driver-gpu`.

Reviewer: ```suggestion
  .Prerequisites                                                                                          
 
  * You have installed {kueue-name} by using the {kueue-op}.
  * You have created a `Kueue` custom resource (CR).                                                      
  * Your cluster is running {product-title} 4.21 or later.
  * A DRA driver is installed in the cluster, for example, `nvidia-dra-driver`. You can verify that the DRA driver is publishing device information by running the following command:
  [source,terminal]
  ----                                                                                                    
  $ oc get resourceslices                                                                                 
  ----
  +                                                                                                       
  If the command returns one or more `ResourceSlice` objects, the DRA driver is running.
  * At least one `DeviceClass` object exists in the cluster. You can verify this by running the following command:
  +                                                                                                       
  [source,terminal]                                                                                       
  ----
  $ oc get deviceclass                                                                                    
  ---- 
```

Author: Added the above changes.

[role="_additional-resources"]
.Additional resources

* xref:../../nodes/pods/nodes-pods-allocate-dra.adoc#nodes-pods-allocate-dra[Allocating GPUs to pods by using DRA]

include::modules/kueue-dra-structured-parameters.adoc[leveloffset=+1]

include::modules/kueue-dra-resourceclaimtemplates.adoc[leveloffset=+2]

> // include::modules/kueue-dra-deviceclasses.adoc[leveloffset=+2]

Reviewer 2: Should this line be commented out?

Author: Yes. it's a file that wasn't needed. I'm removing it from the Assembly file.

:_mod-docs-content-type: ASSEMBLY
include::_attributes/common-attributes.adoc[]
[id="kueue-dra-integrating-dynamic-resource-allocation"]
= Integrating Dynamic Resource Allocation
:context: kueue-dra-integrating-dynamic-resource-allocation

toc::[]

[role="_abstract"]

> {kueue-name} Dynamic Resource Allocation (DRA) integration enables advanced management of specialized hardware resources like GPUs, FPGAs, and other accelerators within Kubernetes workload queuing. This integration allows for publishing ResourceSlices, counter-based quota computation, and specific admission behaviors. 

Reviewer: Clarification on "allows for publishing ResourceSlices"-> Kueue reads ResourceSlices, the DRA driver publishes them.

Author: Replaced with "This integration allows for the reading and publishing of ResourceSlices..."

DRA is a Kubernetes framework that manages specialized hardware resources such as GPUs with fine-grained control. Unlike traditional resource requests, DRA allows dynamic prioritization—allocating GPUs to high-priority AI training workloads during business hours, then reallocating them to cost-optimized batch jobs overnight.

You can validate partitionable devices support in {kueue-name} Dynamic Resource Allocation (DRA) integration, covering partition-aware quota, admission, and scheduling. Partitionable devices, such as NVIDIA MIG, allow graphics processing units (GPUs) to be dynamically subdivided into smaller allocations. {kueue-name} must correctly handle quota accounting for these mutually exclusive partition configurations.

{kueue-name} DRA integration allows for:

* Admission validation with specific partition sizes and quota enforcement.

* Testing of mixed partition workloads on the same device and preemption across partition sizes.

* Verification of partition capacity reclaim after workload completion.

> * Validation using dra-example-driver and nvidia-dra-driver.

Reviewer: `dra-example-driver` is an upstream test fixture, not something OCP users would install. OpenShift docs should reference the supported. NVIDIA DRA driver for OCP, not upstream test tools. Drop dra-example-driver entirely.

Author: updated.

.Prerequisites

> * Kueue is installed.

Reviewer: ```suggestion
* {kueue-name} is installed.
```

Author: changed.

> * A Kubernetes cluster running version 1.34 or later.

Reviewer: It should be `{product-title} 4.21 or later.`

Author: s/A Kubernetes cluster running version 1.34 or later. / {product-title} running version 4.21 or later.

> * A DRA driver installed in the cluster, for example, `dra-example-driver`` for testing, or a vendor driver such as NVIDIA `k8s-dra-driver-gpu` for production.

Reviewer: Drop `dra-example-driver` completely.

Author: Replaced with "A DRA driver installed in the cluster, for example, `nvidia-dra-driver` or `k8s-dra-driver-gpu`."

[role="_additional-resources"]
.Additional resources

* xref:../../nodes/pods/nodes-pods-allocate-dra.adoc#nodes-pods-allocate-dra[Allocating GPUs to pods by using DRA]

include::modules/kueue-dra-structured-parameters.adoc[leveloffset=+1]

include::modules/kueue-dra-resourceclaimtemplates.adoc[leveloffset=+2]

// include::modules/kueue-dra-deviceclasses.adoc[leveloffset=+2]

> include::modules/kueue-dra-extended-resources.adoc[leveloffset=+1]

Reviewer: ER at leveloffset=+1 should be +2 under structured parameters it's an alternative to RCT, not a peer to structured parameters

Author: Changed to `leveloffset=+2`.

> include::modules/kueue-dra-partionable-devices.adoc[leveloffset=+1]

Reviewer: ```suggestion
include::modules/kueue-dra-partitionable-devices.adoc[leveloffset=+1]
```

Author: Fixed.

:_mod-docs-content-type: ASSEMBLY
include::_attributes/common-attributes.adoc[]
[id="kueue-dra-integrating-dynamic-resource-allocation"]
= Integrating Dynamic Resource Allocation
:context: kueue-dra-integrating-dynamic-resource-allocation

toc::[]

> 

Reviewer 3: Per the SSG, the first content in the body text of an assembly or a module should be the short description: https://redhat-documentation.github.io/supplementary-style-guide/#shortdesc

> :FeatureName: Kueue integration with Dynamic Resource Allocation (DRA)

Reviewer 3: ```suggestion

```

> include::snippets/technology-preview.adoc[]

Reviewer 3: ```suggestion

```

[role="_abstract"]
You can configure {kueue-name} to manage quota for workloads that use Dynamic Resource Allocation (DRA) to request GPUs. When DRA quota management is configured, {kueue-name} counts DRA device requests toward quota in the same way that it counts traditional resources such as CPU and memory.

> 

Reviewer 3: ```suggestion

:FeatureName: Kueue integration with Dynamic Resource Allocation (DRA)
include::snippets/technology-preview.adoc[]

```

If DRA device quota is not configured, {kueue-name} does not account for GPU requests when admitting workloads, which can result in teams exceeding their GPU allocation.

> .Prerequisites

Reviewer 3: To meet DITA requirements, we can't include the .Prerequisites heading in an assembly file. I suggest moving the prerequisites to their own module and add an `include` statement to the new module here.

Reviewer 4: I added the prerequisites to the individual modules, edited as needed. I don't think this would make a proper module, as it isn't a true use case. Maybe: 
_What does my cluster need to have in place in addition to the prerequisites in the following modules._

* You have installed {kueue-name} by using the {kueue-op}.
* You have created a `Kueue` custom resource (CR).
* Your cluster is running {product-title} 4.21 or later.
* A DRA driver is installed in the cluster, for example, `nvidia-dra-driver`. You can verify that the DRA driver is publishing device information by running the following command:
+
[source,terminal]
----
$ oc get resourceslices
----
+
If the command returns one or more `ResourceSlice` objects, the DRA driver is running.
* At least one `DeviceClass` object exists in the cluster. You can verify this by running the following command:
+
[source,terminal]
----
$ oc get deviceclass
----

[NOTE]
====

> To use partitionable devices, your cluster must be running {product-title} 4.22 or later and must have the `TechPreviewNoUpgrade` feature set enabled.

Reviewer 5: The [partitionable devices procedure](https://github.com/openshift/openshift-docs/pull/113996/changes#diff-d7fb815ba9aedc31647e85ba1d8729ee1204e4145b1e41b414407aef3c2a4aceR22) instructs users to use CustomNoUpgrade with explicit DRAPartitionableDevices gate enablement, but this note says TechPreviewNoUpgrade. Could we make these consistent?

> To use the extended resources path, you must enable the `DRAExtendedResource` Kubernetes feature gate. This feature is expected to be generally available in a future {product-title} release.

Reviewer 3: Be careful with discussing future releases. SSG: https://redhat-documentation.github.io/supplementary-style-guide/#statements-about-the-future

Reviewer 4: @lahinson  The guidance suggests using _"anticipate", "expect", or "plan"_ and does not mention a specific release. This sentence seems to meet the requirements. Please let me know if you disagree.

Reviewer 3: @mburke5678 Thanks for checking on that. If the guidance says we can use "expected", then the wording here is okay with me.

DRA structured parameters provide quota tracking and  management for specialized hardware such as GPUs, FPGAs, and network adapters. 

You can validate partitionable devices support in {kueue-name} Dynamic Resource Allocation (DRA) integration, covering partition-aware quota, admission, and scheduling. Partitionable devices, such as NVIDIA MIG, allow graphics processing units (GPUs) to be dynamically subdivided into smaller allocations. {kueue-name} must correctly handle quota accounting for these mutually exclusive partition configurations.

{kueue-name} DRA integration allows for:

* Admission validation with specific partition sizes and quota enforcement.

* Testing of mixed partition workloads on the same device and preemption across partition sizes.

* Verification of partition capacity reclaim after workload completion.

* Validation using `nvidia-dra-driver`.

.Prerequisites
* You have installed {kueue-name} by using the {kueue-op}.
* You have created a `Kueue` custom resource (CR).
* Your cluster is running {product-title} 4.21 or later.
+

> ====

Author: fixed.

NOTE: To use partitionable devices, you need {product-title} 4.22 or later. 
====
* A DRA driver installed in the cluster, for example, `nvidia-dra-driver`. You can verify that the DRA driver is publishing device information by running the following command:
+
[source,terminal]
----
$ oc get resourceslices
----
+
If the command returns one or more `ResourceSlice` objects, the DRA driver is running.

* At least one `DeviceClass` object exists in the cluster. You can verify this by running the following command:
+
[source,terminal]
----
$ oc get deviceclass
----

[role="_additional-resources"]
.Additional resources

* xref:../../nodes/pods/nodes-pods-allocate-dra.adoc#nodes-pods-allocate-dra[Allocating GPUs to pods by using DRA]

> 

Reviewer 2: The levels below seem to be not quite correct. I guess we should have something like this:

 = Integrating Dynamic Resource Allocation
    == DRA quota management overview
      === Configuring ResourceClaimTemplates
      === Configuring Extended Resources
      === Configuring Partitionable Devices

 This way the concept module introduces all three, and the procedures sit underneath it.

Author: updated the levels.
